### PR TITLE
Implement --diff-range-notation option

### LINKED ISSUE #105: Wondering why `diff-cover` uses three dot notation for the diff
Hi,

We have integrated `diff-cover` into our workflow at work by making jobs publish `diff-cover` comments into the PR for a branch, including links and emojis (we use BitBucket Server). 

Works great, here's a quick screenshot:

image

So thanks a lot for the great tool!

--- 

Now I would like to describe a problem we are having, and later discuss a couple of possible solutions.

On occasion we work with long lived topic branches. Those are topic branches which will be around for some time (say a month) while a large feature is being worked on. From those long lived topic branches we then create short lived topic branches for smaller amounts of work (for example, a short lived branch to work on the application business logic, then another one after that to work on the GUI, and so on). Then we open PRs to merge the short lived branches into the long lived one.

It happens that we want to get some update from `master` in both the long lived branch and into its short lived branches. Those changes that come from `master` won't appear in the PR from `short topic` -> `long topic`, as BitBucket will only show the changes that are in `short topic` and not on `long topic`, even if they came from somewhere else.

And here we noticed a problem: a few files are being reported as not being fully covered by `diff-cover`, but they are not even shown in our tree of "changed files" on the PR in BitBucket: it is the case with the `test_grid_table_editor.py` in the previous screenshot. This is confusing as we don't really know what to do to fix the coverage, as the file is not even visible to us on the PR.

Digging into the code, I noticed [`diff-cover` uses](https://github.com/Bachmann1234/diff-cover/blob/master/diff_cover/git_diff.py#L33) [three-dot range notation](https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-Theem82308203emthree-dotSymmetricDifferenceNotation) to obtain the patch for the PR.

When using the three dot notation, the commit history is taken in account to produce the patch, including commits which are in both branches but are not in the actual patch that will be applied by the merge.

The other common notation used with `git diff` is [two-dot range notation](https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-Theememtwo-dotRangeNotation), which only shows the full patch between the two branch tips: a plain diff between the trees, regardless of the commit history. 

Another way to explain this:

1. Branch `long` is created from `master`.
2. Branch `short` is created from `big`.
3. A new function, with 0% code coverage, is committed to `master` into commit `ABAB`.
4. `master` is merged into `long` **and** `short`, so both branches now contain the commit `ABAB`.

With two-dot notation, the code coverage for the `short -> long` merge shows nothing (that's the exact patch between them).

With three-dot notation, the code coverage for the `short -> long` merge will show the function with 0% coverage, even thought it was not added by the developers of either `long` or `short` branches, coming from `master` (of course it should have been covered in the first place, but that's beside the point).

IMHO the two-dot notation makes more sense for `diff-cover`, this way developers only worry about the actual lines being changed by them (which codecov.io calls "patch coverage"), which I believe is one of the selling points of using `diff-cover`. :grin:

---

Having said all that (and sorry about the long post), I see some solutions:

1. Change `diff-cover` to use two-dot notation. This is simplest, but I'm not sure how workflow works, and it would break use cases where the `diff-cover` notation makes sense.
2. Make a `--diff-notation=two/three` (or similar) option that lets the (advanced) user choose which method to use to obtain the patch.
3. Restart the work in #26 and let the user provide a patch file, bypassing the actual call to `git diff`.

What do you guys think? I'm happy to contribute a PR with whatever solution you deem best.

Thanks!

### PR DESCRIPTION
**Pull request by nicoddemus**
I've also used `self.addCleanup` to undo setup changes in a few places, after noticing that the patches in `test_git_diff.py` were not being undone.

`addCleanup` is considered more resilient and safer because `tearDown` methods are not executed in case the `setUp` method fails, which might leave resources around.

They also bring resource creation and deletion closer together, facilitating refactorings.

Fix #105

### CODE REVIEW

File: diff_cover/diff_cover_tool.py [hunk 1]
```diff
@@ -116,19 +117,28 @@ def parse_coverage_args(argv):
         help=SRC_ROOTS_HELP
     )
 
+    parser.add_argument(
+        '--diff-range-notation',
+        metavar='RANGE_NOTATION',
+        type=str,
+        default='...',
```
Comment (Bachmann1234) [inline diff_cover/diff_cover_tool.py]: So wait, I thought there was only double dot and triple dot. Are there more? 

If there only the two perhaps we just have a flag like `--double-dot-diff-notation` or something like that? Im open to thoughts. Mostly just talking it out.

**Comment by Bachmann1234** [comment]
Looks good to me. I left a question and curious on your thinking. But if you feel strongly about leaving it as is im happy to leave it and merge

File: diff_cover/diff_cover_tool.py (hunk 1 above)
Comment (nicoddemus) [reply diff_cover/diff_cover_tool.py]: Those are the most common. `--double-dot-diff-notation` might be simpler, although it doesn't make it clear what *else* is done if `--double-dot-diff-notation` is passed. On the other hand `--diff-range-notation` being either `...` or `..` makes that more explicit.

Having said that I don't mind changing it. Let me know what you think.

Comment (Bachmann1234) [reply diff_cover/diff_cover_tool.py]: You make some good points. I think im ok with it but can you add a small validation function? Even as something as using the `choices` argument to argparse. It would restrict users to `...` and `..` but like... if someone wants another they can open a pr/issue. (If you know of another potentially good one feel free but like... I think starting with what has an immediate need has served the project well to this point) 

Basically I would really want to have an error thrown before some invalid string got sent to a shell.

Comment (nicoddemus) [reply diff_cover/diff_cover_tool.py]: Done! I also rebased on `master`.

**Comment by nicoddemus** [comment]
After you are happy with this changes and this gets merged, could we have another release? I'm happy to open the PR preparing the release if that helps. I don't see instructions on how to make a release, but from what I can gather:

* Update version in `__init__.py`
* Update CHANGELOG
* Open a PR; after the PR gets approved, someone with write access should push a tag to the PR's HEAD and then finally merge it.

**Comment by Bachmann1234** [comment]
Woot! Thanks for all the work on this. Ill handle getting the release out. let me look at one other PR and get this done tonight

**Comment by nicoddemus** [comment]
Awesome, thanks!

**Comment by Bachmann1234** [comment]
just crossing my fingers hoping my recently added 2fa did not break the travis CD pipeline....

**Comment by Bachmann1234** [comment]
woot! https://pypi.org/project/diff-cover/2.1.0/

**Comment by nicoddemus** [comment]
Thanks!
